File: packages/python-sdk/onesub/api/credits.py
"""Credits API"""

import time
import random
from typing import Any, Dict, Optional, Tuple, Union
from ..http import HttpClient
from ..errors import ValidationError, InsufficientCreditsError


class CreditsApi:
    """Credits API client"""

    def __init__(self, http: HttpClient):
        self._http = http

    def consume(
        self,
        user_id: str,
        amount: int,
        reason: str,
        idempotency_key: str
    ) -> Dict[str, Any]:
        """
        Consume credits from a user's balance.

        Args:
            user_id: The user's 1Sub ID
            amount: Amount of credits to consume (1 - 1,000,000)
            reason: Reason for credit consumption
            idempotency_key: Unique key for safe retries

        Returns:
            Dict with success, new_balance, transaction_id, is_duplicate

        Raises:
            InsufficientCreditsError: When user doesn't have enough credits
        """
        # Validate inputs
        if not user_id:
            raise ValidationError("user_id must be provided")
        if not isinstance(amount, int) or amount <= 0:
            raise ValidationError("amount must be a positive integer")
        if amount > 1_000_000:
            raise ValidationError("amount cannot exceed 1,000,000")
        if not reason:
            raise ValidationError("reason must be provided")
        if len(reason) > 500:
            raise ValidationError("reason cannot exceed 500 characters")
        if not idempotency_key:
            raise ValidationError("idempotency_key must be provided")
        if len(idempotency_key) > 255:
            raise ValidationError("idempotency_key cannot exceed 255 characters")

        data = self._http.post("/credits/consume", {
            "user_id": user_id,
            "amount": amount,
            "reason": reason,
            "idempotency_key": idempotency_key,
        })

        return {
            "success": data.get("success", True),
            "new_balance": data.get("new_balance"),
            "transaction_id": data.get("transaction_id"),
            "is_duplicate": data.get("is_duplicate", False),
        }

    def try_consume(
        self,
        user_id: str,
        amount: int,
        reason: str,
        idempotency_key: str
    ) -> Union[Tuple[bool, Dict[str, Any]], Tuple[bool, str, Optional[int]]]:
        """
        Try to consume credits, returning success/failure.

        Returns:
            (True, data) on success
            (False, error_message, shortfall) on failure
        """
        try:
            data = self.consume(user_id, amount, reason, idempotency_key)
            return (True, data)
        except InsufficientCreditsError as e:
            return (False, e.message, e.shortfall)
        except Exception as e:
            return (False, str(e), None)

    def has_enough(self, user_id: str, amount: int) -> bool:
        """
        Check if user has enough credits.
        Note: Balance may change between check and consume.
        """
        try:
            data = self._http.post("/tools/subscriptions/verify", {
                "oneSubUserId": user_id
            })
            balance = data.get("creditsRemaining", 0)
            return balance >= amount
        except Exception:
            return False

    @staticmethod
    def generate_idempotency_key(*parts: str) -> str:
        """Generate a unique idempotency key"""
        timestamp = hex(int(time.time()))[2:]
        rand = hex(random.randint(0, 0xFFFFFF))[2:]
        return "-".join(list(parts) + [timestamp, rand])

File: packages/sdk/src/api/credits.ts
import type {
  ConsumeCreditsRequest,
  ConsumeCreditsResponse,
} from '../types.js';
import { HttpClient } from '../utils/http.js';
import { ValidationError } from '../utils/errors.js';

/**
 * Credits API client
 */
export class CreditsApi {
  private readonly http: HttpClient;

  constructor(http: HttpClient) {
    this.http = http;
  }

  /**
   * Consume credits from a user's balance
   *
   * @example
   * ```typescript
   * const result = await onesub.credits.consume({
   *   userId: 'uuid-123',
   *   amount: 10,
   *   reason: 'Image generation',
   *   idempotencyKey: 'req-abc-123'
   * });
   *
   * console.log(`New balance: ${result.newBalance}`);
   * ```
   *
   * @throws {InsufficientCreditsError} When user doesn't have enough credits
   */
  async consume(params: ConsumeCreditsRequest): Promise<ConsumeCreditsResponse> {
    // Validate required fields
    if (!params.userId || typeof params.userId !== 'string') {
      throw new ValidationError('userId must be a non-empty string');
    }

    if (typeof params.amount !== 'number' || params.amount <= 0) {
      throw new ValidationError('amount must be a positive number');
    }

    if (params.amount > 1000000) {
      throw new ValidationError('amount cannot exceed 1,000,000');
    }

    if (!params.reason || typeof params.reason !== 'string') {
      throw new ValidationError('reason must be a non-empty string');
    }

    if (params.reason.length > 500) {
      throw new ValidationError('reason cannot exceed 500 characters');
    }

    if (!params.idempotencyKey || typeof params.idempotencyKey !== 'string') {
      throw new ValidationError('idempotencyKey must be a non-empty string');
    }

    if (params.idempotencyKey.length > 255) {
      throw new ValidationError('idempotencyKey cannot exceed 255 characters');
    }

    // Make API request
    const response = await this.http.post<{
      success: boolean;
      new_balance: number;
      transaction_id: string;
      is_duplicate?: boolean;
    }>('/credits/consume', {
      user_id: params.userId,
      amount: params.amount,
      reason: params.reason,
      idempotency_key: params.idempotencyKey,
    });

    // Transform snake_case response to camelCase
    return {
      success: response.data.success,
      newBalance: response.data.new_balance,
      transactionId: response.data.transaction_id,
      isDuplicate: response.data.is_duplicate,
    };
  }

  /**
   * Try to consume credits, returning success/failure instead of throwing
   *
   * @example
   * ```typescript
   * const result = await onesub.credits.tryConsume({
   *   userId: 'uuid-123',
   *   amount: 10,
   *   reason: 'Image generation',
   *   idempotencyKey: 'req-abc-123'
   * });
   *
   * if (result.success) {
   *   console.log(`Credits consumed, new balance: ${result.newBalance}`);
   * } else {
   *   console.log(`Failed: ${result.error}`);
   * }
   * ```
   */
  async tryConsume(
    params: ConsumeCreditsRequest
  ): Promise<
    | { success: true; data: ConsumeCreditsResponse }
    | { success: false; error: string; shortfall?: number }
  > {
    try {
      const data = await this.consume(params);
      return { success: true, data };
    } catch (error) {
      if (error instanceof Error) {
        // Check if it's an insufficient credits error
        if ('shortfall' in error) {
          return {
            success: false,
            error: error.message,
            shortfall: (error as { shortfall: number }).shortfall,
          };
        }
        return { success: false, error: error.message };
      }
      return { success: false, error: 'Unknown error' };
    }
  }

  /**
   * Check if user has enough credits for an operation
   * Note: This doesn't reserve credits, so balance may change between check and consume
   *
   * @example
   * ```typescript
   * const hasCredits = await onesub.credits.hasEnough('uuid-123', 100);
   * if (hasCredits) {
   *   // Proceed with operation
   * }
   * ```
   */
  async hasEnough(userId: string, amount: number): Promise<boolean> {
    // NOTE: This method requires a verification token, which is not available in this context
    // It's recommended to check credits from the entitlements returned by the verify endpoint
    // This is a workaround that will be deprecated
    console.warn('hasEnough() is deprecated. Check credits from verify() entitlements instead.');

    // Return false as we can't check without a verification token in the new flow
    return false;
  }

  /**
   * Generate a unique idempotency key
   * Useful for creating request-specific keys
   *
   * @example
   * ```typescript
   * const key = onesub.credits.generateIdempotencyKey('image-gen', userId, requestId);
   * ```
   */
  generateIdempotencyKey(...parts: string[]): string {
    const timestamp = Date.now().toString(36);
    const random = Math.random().toString(36).substring(2, 8);
    return [...parts, timestamp, random].join('-');
  }
}
